x/bsonx/bsoncore/array_test.go

```diff
@@ -0,0 +1,239 @@
+// Copyright (C) MongoDB, Inc. 2017-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package bsoncore
+
+import (
+	"bytes"
+	"encoding/binary"
+	"io"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+func TestArray(t *testing.T) {
+	t.Run("Validate", func(t *testing.T) {
+		t.Run("TooShort", func(t *testing.T) {
+			want := NewInsufficientBytesError(nil, nil)
+			got := Array{'\x00', '\x00'}.Validate()
+			if !compareErrors(got, want) {
+				t.Errorf("Did not get expected error. got %v; want %v", got, want)
+			}
+		})
+		t.Run("InvalidLength", func(t *testing.T) {
+			want := Array{}.lengtherror(200, 5)
+			r := make(Array, 5)
+			binary.LittleEndian.PutUint32(r[0:4], 200)
+			got := r.Validate()
+			if !compareErrors(got, want) {
+				t.Errorf("Did not get expected error. got %v; want %v", got, want)
+			}
+		})
+		t.Run("Invalid Element", func(t *testing.T) {
+			want := NewInsufficientBytesError(nil, nil)
+			r := make(Array, 9)
+			binary.LittleEndian.PutUint32(r[0:4], 9)
+			r[4], r[5], r[6], r[7], r[8] = 0x02, 'f', 'o', 'o', 0x00
+			got := r.Validate()
+			if !compareErrors(got, want) {
+				t.Errorf("Did not get expected error. got %v; want %v", got, want)
+			}
+		})
+		t.Run("Missing Null Terminator", func(t *testing.T) {
+			want := ErrMissingNull
+			r := make(Array, 8)
+			binary.LittleEndian.PutUint32(r[0:4], 8)
+			r[4], r[5], r[6], r[7] = 0x0A, 'f', 'o', 'o'
+			got := r.Validate()
+			if !compareErrors(got, want) {
+				t.Errorf("Did not get expected error. got %v; want %v", got, want)
+			}
+		})
+		testCases := []struct {
+			name string
+			r    Array
+			want error
+		}{
+			{"null", Array{'\x08', '\x00', '\x00', '\x00', '\x0A', 'x', '\x00', '\x00'}, nil},
+			{"subarray",
+				Array{
```

Comment:
The bytes in this Array are the ones in the corresponding `Document` test, so we're not actually testing sub-arrays. Printing this array shows `["foo": {"a": null,"b": null}]`, which isn't actually a valid array at all.

Reply:
You're right... Which, as discussed offline, means something is wrong with `bsoncore.Array#Validate()`. I'm still looking into it, as I don't understand the exact semantics of BSON representations of arrays.

Reply:
http://bsonspec.org/spec.html notes the key restriction for an array type:

> Array - The document for an array is a normal BSON document with integer values for the keys, starting with 0 and continuing sequentially. For example, the array ['red', 'blue'] would be encoded as the document {'0': 'red', '1': 'blue'}. The keys must be in ascending numerical order.

I think you could keep this test, and have `bsoncore.Array#Validate()` additionally check `Element#Key()`, and update this test and the one below to expect an error, and add one case that should pass.

Aside: I do not find reading or writing raw BSON fun. One tool I've found helpful before this script David Golden wrote: 
https://github.com/mongodb/specifications/blob/master/source/bson-corpus/tests/bsonview

E.g. if you dump this byte slice as hex:
```
fmt.Printf("%x\n", []byte(tc.r))
// prints 1500000003666f6f000b0000000a61000a62000000
```

You can echo that into bsonview to get color-coded output, with spacing between lengths and subtypes:
```
> echo "1500000003666f6f000b0000000a61000a62000000" | ~/code/specifications/source/bson-corpus/tests/bsonview -x
15000000 03 "foo" 00 { 0b000000 0a "a" 00 0a "b" 00 00 } 00
```

Reply:
Thanks for all the info! `bsonview` was super helpful 🎉 .

I've updated `bsoncore.Array#Validate()` to check the keys to make sure they begin at "0" and increase sequentially (this works on embedded arrays, too). I've updated these tests to use BSON representations of arrays, and added a couple error cases (keys not increasing, keys not numbers).